src/server/pkg/storage/chunk/reader.go

```diff
@@ -104,6 +124,77 @@ func (r *Reader) readChunk(chunk *Chunk) error {
 	return nil
 }
 
+// AtSplit registers a callback for when a chunk split point is encountered.
+// The callback is only executed at a split point found after reading WindowSize bytes.
+// The reason for this is to guarantee that the same split point will appear in the writer
+// the data is being written to.
+func (r *Reader) AtSplit(f func()) {
+	r.bytesBeforeSplit = 0
+	r.atSplit = f
+}
+
+func (r *Reader) executeAtSplitFunc() {
+	if r.atSplit != nil && r.bytesBeforeSplit > WindowSize {
+		r.atSplit()
+		r.atSplit = nil
+	}
+}
+
+// Copy is the basic data structure to represent a copy of data from
+// a reader to a writer.
+type Copy struct {
+	before, after *bytes.Buffer
+	chunkRefs     []*DataRef
+}
+
+// ReadCopy reads copy data from the reader.
+func (r *Reader) ReadCopy(n ...int64) (*Copy, error) {
```

Comment:
It seems a little weird to me that `Copy` is a public struct that's returned by a public method and yet it doesn't have any public fields. What is a caller of `ReadCopy` supposed to do with the `Copy` they get back?

Reply:
The caller would just store it to be shuffled during the copy process. The package the copy struct is associated with (chunk/index) knows how to use it Remember the path for copying is reader (chunk -> index -> fileset) -> writer (fileset -> index -> chunk). This is just a way to encapsulate the details of copying and not require each layer of abstraction to reach into the reader/writer for lower level abstractions.

Reply:
I see, so it's an opaque structure that gets passed around between methods? It seems like if all you can do is get it out and pass it to other methods it could just be a private method, and those methods that take copies could just take a `Reader` instead and call the private methods. I'm generally a fan of as few exposes primitives as possible.

Reply:
I could potentially do something like this, but it would not be as simple as just calling a copy function attached to a reader from the corresponding writer because the higher level reader abstraction is responsible for setting the bounds on the copy. I will make a note to re-visit this idea down the line (before we ship).